connectors/services/job_execution.py

```diff
@@ -0,0 +1,146 @@
+#
+from functools import cached_property
+
+from connectors.es.client import License
+from connectors.es.index import DocumentNotFoundError
+from connectors.es.license import requires_platinum_license
+from connectors.logger import logger
+from connectors.protocol import (
+    ConnectorIndex,
+    DataSourceError,
+    JobType,
+    SyncJobIndex,
+)
+from connectors.services.base import BaseService
+from connectors.source import get_source_klass
+from connectors.sync_job_runner import SyncJobRunner
+from connectors.utils import ConcurrentTasks
+
+
+class JobExecutionService(BaseService):
+    name = "execute"
+    display_name = "job execution"
+    JOB_TYPES = [
+        JobType.ACCESS_CONTROL.value,
+        JobType.FULL.value,
+        JobType.INCREMENTAL.value,
+    ]
+
+    def __init__(self, config):
+        super().__init__(config)
+        self.idling = self.service_config["idling"]
+        self.source_list = config["sources"]
+        self.sync_job_pool = ConcurrentTasks(max_concurrency=self.max_concurrency)
+
+    def stop(self):
+        super().stop()
+        self.sync_job_pool.cancel()
+
+    @cached_property
+    def max_concurrency(self):
+        raise NotImplementedError()
+
+    def should_execute(self, connector, sync_job):
+        return False
+
+    async def _sync(self, sync_job):
+        if sync_job.service_type not in self.source_list:
+            msg = f"Couldn't find data source class for {sync_job.service_type}"
+            raise DataSourceError(msg)
+        source_klass = get_source_klass(self.source_list[sync_job.service_type])
+        connector_id = sync_job.connector_id
+
+        try:
+            connector = await self.connector_index.fetch_by_id(connector_id)
+        except DocumentNotFoundError:
+            sync_job.log_error("Couldn't find connector")
+            return
+
+        if requires_platinum_license(sync_job, connector, source_klass):
+            (
+                is_platinum_license_enabled,
+                license_enabled,
+            ) = await self.connector_index.has_active_license_enabled(License.PLATINUM)
+
+            if not is_platinum_license_enabled:
+                sync_job.log_error(
+                    f"Minimum required Elasticsearch license: '{License.PLATINUM.value}'. Actual license: '{license_enabled.value}'."
+                )
+                return
+
+        if not self.should_execute(connector, sync_job):
+            return
+
+        sync_job_runner = SyncJobRunner(
+            source_klass=source_klass,
+            sync_job=sync_job,
+            connector=connector,
+            es_config=self._override_es_config(connector),
+        )
+        if not self.sync_job_pool.try_put(sync_job_runner.execute):
+            sync_job.log_warning(
+                f"{self.display_name.capitalize()} has reached the maximum concurrency of {self.max_concurrency}. Retry in {self.idling} seconds."
+            )
```

Comment:
I think this message can be improved to indicate some points to the user:

1. What's up with "maximum concurrency" in this case? We can say that service X is already running Y jobs and cannot run more at this point.
2. What's the action user can take? For example, increase "Z" setting in the connector framework to allow more jobs to run.
3. Should it be a warning or info? I can imagine this being spammed every 30 seconds if jobs take too long to finish, and warnings that often look scary. To me this logic is instead a part of normal flow, so info/debug might suit better

Reply:
>We can say that service X is already running Y jobs and cannot run more at this point.

Will rephrase it, although I think they are more or less the same 😆 

>What's the action user can take? For example, increase "Z" setting in the connector framework to allow more jobs to run.

That's a good point, I will add this to the log message.

>To me this logic is instead a part of normal flow, so info/debug might suit better

👍

Reply:
Updated in https://github.com/elastic/connectors/pull/1865/commits/b87c970c16f23abfb6aeb9a6297ecc893f67f178

The message is like:

>Content sync job execution service is already running 4 sync jobs and can't run more at this poinit. It will retry in 30 seconds. Increase 'service.max_concurrent_content_syncs' in config if you want the service to run more sync jobs.


>Access control sync job execution service is already running 4 sync jobs and can't run more at this poinit. It will retry in 30 seconds. Increase 'service.max_concurrent_access_control_syncs' in config if you want the service to run more sync jobs.